File: java/java-core-learning-example/src/main/java/org/arithmetic/BubbleSort.java
package org.arithmetic;

public class BubbleSort {

    /*  冒泡排序
        两两相邻比较
        然后排序
     */
    public void sort(int[] array) {
        // 创建一个数组
        for (int i = 1; i < array.length; i++) {
            for (int j = 0; j < array.length; j++) {
                if (array[j] > array[j + 1]) {
                    int temp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = temp;
                }
            }
        }
        for (int i : array) {
            System.out.println(">" + i);
        }
    }

    public static void main(String[] args) {
        int[] array = {63, 4, 24, 1, 3, 15};
        BubbleSort sorter = new BubbleSort();
        sorter.sort(array);
    }
}

File: alg/sort/bubble_sort.go
package sort

/**
  排序思想；   又称分区交换排序,
  时间复杂读:  最糟糕的情况是 O(n2) 平均是 O(log n)
  空间复杂度:  O(n) 平均是 O(log n)
*/

func bubbleSort(nums []int) []int {
	hasChange := true

	for i := 0; i < len(nums)-1 && hasChange; i++ {
		hasChange = false
		for j := 0; j < len(nums)-1; j++ {
			if nums[j] > nums[j+1] {
				nums[j+1], nums[j] = nums[j], nums[j+1]
				hasChange = true
			}
		}
	}
	return nums
}
